nodestream/pipeline/pipeline.py

```diff
@@ -78,9 +145,49 @@ async def work_body(self):
                 self.progress_reporter.report(index, record)
 
     async def work_loop(self):
```

Comment:
This function has gotten very complex now. I think we have an opportunity to have this function tell a higher-level story and then have helper functions handle the details. With the refactoring made from the previous review of this PR, i think this can be as simple as a `work_loop` function that looks like this:

```python
async def work_loop(self):
   self.try_start()
   await self.try_work_body()
   await self.try_stop()
   self.raise_encountered_exceptions()
```